src/meili/map_matcher.cc

```diff
@@ -318,12 +315,11 @@ MatchResult FindMatchResult(const MapMatcher& mapmatcher,
     if (rbegin == rend)
       break;
     while (rbegin != rend) {
-      // the very first state will have an initial dummy label with invalid edge/node ids
       const auto& label = *rbegin;
-      if (!label.edgeid().Is_Valid() && !label.nodeid().Is_Valid())
-        break;
-      next_edge = label.edgeid();
-      node_id = label.nodeid();
+      if (label.nodeid().Is_Valid())
```

Comment:
I don't understand why it has to iterate the whole route here. Previously it just returned the rbegin's end location (rbegin->edgeid() and rbegin->target). But if it has to iterate the whole route then check how MergeRoute does:
```c++
  for (; std::next(rbegin) != rend; rbegin++) {
    // nodeid might be invalid if the label ends at an destination on the edge
    if (rbegin->nodeid().IsValid()) {node_id = rbegin->nodeid();}
    // edgeid is always valid
    assert rbegin->edgeid()->IsValid();
    next_edge = rbegin->edgeid();
  }
```

Reply:
@ptpt the larger comment blocks above try to hint about why this is the case. the short answer is, because of node routing and if `interpolation_distance` is `0` you can get the case where you have situation where there are two adjacent points in the input which both snapped to the same graph node. in this case the routing will return a labelset with just one dummy node label representing the trivial node to node route where you route from one node to the same node haha. the problem with that is that this function needs to return an edge id for the part of the path that this match result matched to. we cant do that with a nodeid, so what we do is we iterate forward or backward in states so that we can find the actual path that connected to the node and use that edge id. its like the comments say above, its rare but we need to do it in special cases of trivial node to node routes

Reply:
@kevinkreiser Aha I see, that's indeed a unfortunate case. In this case there is no way to find the match edge, right? It is like another unfortunate case: map match a single point that snaps to a node -- all we could do is just take one of the edges that starting from the node as the match?

Let me go through `FindMatchResult` with this in mind first and see how it's handled and what I missed.

Reply:
Yep, I believe the suggested change works for all cases, the node to same node case and it also handles traversing all the labels back to the original dummy label because inside of the `routepathiterator` we are looking for invalid predecessor index. we could detect the dummy label if we wanted to and stop one iteration earlier in the case of node routes but i think this one more iteration doesnt matter. i like the simplicity of the changes proposed here. though to your comment below i do think a `for` loop would be nicer than a `while` loop if others are so inclined